File: app/src/main/java/org/lineageos/recorder/RecorderActivity.java
/*
 * Copyright (C) 2017-2022 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.recorder;

import static android.provider.MediaStore.Audio.Media.RECORD_SOUND_ACTION;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.telephony.TelephonyManager;
import android.text.format.DateUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.lineageos.recorder.service.SoundRecorderService;
import org.lineageos.recorder.status.UiStatus;
import org.lineageos.recorder.task.DeleteRecordingTask;
import org.lineageos.recorder.task.TaskExecutor;
import org.lineageos.recorder.ui.WaveFormView;
import org.lineageos.recorder.utils.AppPreferences;
import org.lineageos.recorder.utils.LocationHelper;
import org.lineageos.recorder.utils.OnBoardingHelper;
import org.lineageos.recorder.utils.PermissionManager;
import org.lineageos.recorder.utils.Utils;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

public class RecorderActivity extends AppCompatActivity {
    private static final String FILE_NAME_BASE = "SoundRecords/%1$s (%2$s)";
    private static final String FILE_NAME_FALLBACK = "Sound record";

    private FloatingActionButton mSoundFab;
    private ImageView mPauseResume;

    private TextView mRecordingText;
    private TextView mElapsedTimeText;
    private WaveFormView mRecordingVisualizer;

    private LocationHelper mLocationHelper;
    private AppPreferences mPreferences;
    private PermissionManager mPermissionManager;
    private TaskExecutor mTaskExecutor;

    private boolean mReturnAudio;
    private boolean mHasRecordedAudio;

    private final StringBuilder mSbRecycle = new StringBuilder();

    @UiStatus
    private int mUiStatus = UiStatus.READY;

    private final BroadcastReceiver mTelephonyReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (TelephonyManager.ACTION_PHONE_STATE_CHANGED.equals(intent.getAction())) {
                String state = intent.getStringExtra(TelephonyManager.EXTRA_STATE);
                if (TelephonyManager.EXTRA_STATE_OFFHOOK.equals(state)) {
                    togglePause();
                }
            }
        }
    };

    private final Messenger mMessenger = new Messenger(new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(@NonNull Message msg) {
            switch (msg.what) {
                case SoundRecorderService.MSG_UI_STATUS:
                    setStatus(msg.arg1);
                    break;
                case SoundRecorderService.MSG_SOUND_AMPLITUDE:
                    setVisualizerAmplitude(msg.arg1);
                    break;
                case SoundRecorderService.MSG_TIME_ELAPSED:
                    // elapsed time is a long value split in two integer values
                    setElapsedTime(((long) msg.arg1 << 32) | msg.arg2 & 0xffffffffL);
                    break;
                default:
                    super.handleMessage(msg);
                    break;
            }
        }
    });
    private final ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mService = new Messenger(service);

            try {
                final Message msg = Message.obtain(null,
                        SoundRecorderService.MSG_REGISTER_CLIENT);
                msg.replyTo = mMessenger;
                mService.send(msg);
            } catch (RemoteException e) {
                // In this case the service has crashed before we could even
                // do anything with it; we can count on soon being
                // disconnected (and then reconnected if it can be restarted)
                // so there is no need to do anything here.
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mService = null;
        }
    };
    private Messenger mService;

    @Override
    public void onCreate(@Nullable Bundle savedInstance) {
        super.onCreate(savedInstance);
        setContentView(R.layout.activity_main);

        ConstraintLayout mainView = findViewById(R.id.main_root);
        mSoundFab = findViewById(R.id.sound_fab);
        mPauseResume = findViewById(R.id.sound_pause_resume);
        ImageView soundList = findViewById(R.id.sound_list_icon);
        ImageView settings = findViewById(R.id.sound_settings);

        mRecordingText = findViewById(R.id.main_title);
        mElapsedTimeText = findViewById(R.id.main_elapsed_time);
        mRecordingVisualizer = findViewById(R.id.main_recording_visualizer);

        mSoundFab.setOnClickListener(v -> toggleSoundRecorder());
        mPauseResume.setOnClickListener(v -> togglePause());
        soundList.setOnClickListener(v -> openList());
        settings.setOnClickListener(v -> openSettings());

        Utils.setFullScreen(getWindow(), mainView);
        Utils.setVerticalInsets(mainView);

        mLocationHelper = new LocationHelper(this);
        mPreferences = AppPreferences.getInstance(this);
        mPermissionManager = new PermissionManager(this);

        mTaskExecutor = new TaskExecutor();
        getLifecycle().addObserver(mTaskExecutor);

        if (RECORD_SOUND_ACTION.equals(getIntent().getAction())) {
            mReturnAudio = true;
            soundList.setVisibility(View.GONE);
            settings.setVisibility(View.GONE);
        }

        doBindService();

        OnBoardingHelper.onBoardList(this, soundList);
        OnBoardingHelper.onBoardSettings(this, settings);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        doUnbindService();
    }

    @Override
    protected void onStart() {
        super.onStart();
        registerReceiver(mTelephonyReceiver,
                new IntentFilter(TelephonyManager.ACTION_PHONE_STATE_CHANGED));
    }

    @Override
    protected void onStop() {
        super.onStop();
        unregisterReceiver(mTelephonyReceiver);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] results) {
        super.onRequestPermissionsResult(requestCode, permissions, results);
        if (requestCode == PermissionManager.REQUEST_CODE) {
            if (mPermissionManager.hasEssentialPermissions()) {
                toggleAfterPermissionRequest();
            } else {
                mPermissionManager.onEssentialPermissionsDenied();
            }
        }
    }

    private void setStatus(@UiStatus int status) {
        applyUiStatus(status);

        if (mReturnAudio && mHasRecordedAudio) {
            Utils.cancelShareNotification(this);
            promptUser();
        }
    }

    private void setVisualizerAmplitude(int amplitude) {
        mRecordingVisualizer.post(() -> mRecordingVisualizer.setAmplitude(amplitude));
    }

    private void setElapsedTime(long seconds) {
        mElapsedTimeText.post(() -> mElapsedTimeText.setText(
                DateUtils.formatElapsedTime(mSbRecycle, seconds)));
    }

    private void toggleAfterPermissionRequest() {
        doBindService();
        new Handler(Looper.getMainLooper()).postDelayed(this::toggleSoundRecorder, 500);
    }

    private void toggleSoundRecorder() {
        if (mPermissionManager.requestEssentialPermissions()) {
            return;
        }

        if (mConnection == null) {
            doBindService();
            return;
        }

        if (mUiStatus == UiStatus.READY) {
            // Start
            startService(new Intent(this, SoundRecorderService.class)
                    .setAction(SoundRecorderService.ACTION_START)
                    .putExtra(SoundRecorderService.EXTRA_FILE_NAME, getNewRecordFileName()));
        } else {
            // Stop
            startService(new Intent(this, SoundRecorderService.class)
                    .setAction(SoundRecorderService.ACTION_STOP));
            mHasRecordedAudio = true;
        }
    }

    private void togglePause() {
        switch (mUiStatus) {
            case UiStatus.RECORDING:
                startService(new Intent(this, SoundRecorderService.class)
                        .setAction(SoundRecorderService.ACTION_PAUSE));
                break;
            case UiStatus.PAUSED:
                startService(new Intent(this, SoundRecorderService.class)
                        .setAction(SoundRecorderService.ACTION_RESUME));
                break;
            case UiStatus.READY:
                // Do nothing
                break;
        }
    }

    private void applyUiStatus(@UiStatus int status) {
        mUiStatus = status;

        if (UiStatus.READY == status) {
            mRecordingText.setText(getString(R.string.main_sound_action));
            mSoundFab.setImageResource(R.drawable.ic_action_record);
            mElapsedTimeText.setVisibility(View.GONE);
            mRecordingVisualizer.setVisibility(View.GONE);
            mPauseResume.setVisibility(View.GONE);
        } else {
            mSoundFab.setImageResource(R.drawable.ic_action_stop);
            mElapsedTimeText.setVisibility(View.VISIBLE);
            mRecordingVisualizer.setVisibility(View.VISIBLE);
            mRecordingVisualizer.setAmplitude(0);
            mPauseResume.setVisibility(View.VISIBLE);
            final Drawable prDrawable;
            if (UiStatus.PAUSED == status) {
                mRecordingText.setText(getString(R.string.sound_recording_title_paused));
                mPauseResume.setContentDescription(getString(R.string.resume));
                prDrawable = ContextCompat.getDrawable(this, R.drawable.avd_play_to_pause);
            } else {
                mRecordingText.setText(getString(R.string.sound_recording_title_working));
                mPauseResume.setContentDescription(getString(R.string.pause));
                prDrawable = ContextCompat.getDrawable(this, R.drawable.avd_pause_to_play);
            }
            mPauseResume.setTooltipText(mPauseResume.getContentDescription());
            mPauseResume.setImageDrawable(prDrawable);
            if (prDrawable instanceof AnimatedVectorDrawable) {
                ((AnimatedVectorDrawable) prDrawable).start();
            }
        }
    }

    private void doBindService() {
        if (mPermissionManager.hasEssentialPermissions()) {
            bindService(new Intent(this, SoundRecorderService.class),
                    mConnection, BIND_AUTO_CREATE);
        }
    }

    private void doUnbindService() {
        if (mService != null) {
            try {
                final Message msg = Message.obtain(null,
                        SoundRecorderService.MSG_UNREGISTER_CLIENT);
                msg.replyTo = mMessenger;
                mService.send(msg);
            } catch (RemoteException e) {
                // There is nothing special we need to do if the service
                // has crashed.
            }
        }
        unbindService(mConnection);
    }

    private void openSettings() {
        Intent intent = new Intent(this, DialogActivity.class)
                .putExtra(DialogActivity.EXTRA_IS_RECORDING, mUiStatus != UiStatus.READY);
        startActivity(intent);
    }

    private void openList() {
        startActivity(new Intent(this, ListActivity.class));
    }

    private void confirmLastResult() {
        Intent resultIntent = new Intent().setData(mPreferences.getLastItemUri());
        setResult(RESULT_OK, resultIntent);
        finish();
    }

    private void discardLastResult() {
        final Uri uri = mPreferences.getLastItemUri();
        if (uri != null) {
            mTaskExecutor.runTask(new DeleteRecordingTask(getContentResolver(), uri), () -> {
                Utils.cancelShareNotification(this);
                mPreferences.setLastItemUri(null);
            });
        }
        cancelResult(true);
    }

    private void cancelResult(boolean quit) {
        setResult(RESULT_CANCELED, new Intent());
        mHasRecordedAudio = false;
        if (quit) {
            finish();
        }
    }

    private void promptUser() {
        new AlertDialog.Builder(this)
                .setTitle(R.string.confirm_result_title)
                .setMessage(R.string.confirm_result_message)
                .setPositiveButton(R.string.confirm, (dialog, which) -> confirmLastResult())
                .setNegativeButton(R.string.discard, (dialog, which) -> discardLastResult())
                .setNeutralButton(R.string.record_again, (dialog, which) -> cancelResult(false))
                .setCancelable(false)
                .show();
    }

    private String getNewRecordFileName() {
        final String tag = mLocationHelper.getCurrentLocationName()
                .orElse(FILE_NAME_FALLBACK);
        final DateTimeFormatter formatter = DateTimeFormatter.ofPattern(
                getString(R.string.main_file_date_time_format),
                Locale.getDefault());
        return String.format(FILE_NAME_BASE, tag,
                formatter.format(LocalDateTime.now())) + ".%1$s";
    }
}

File: app/src/main/java/org/lineageos/recorder/RecorderActivity.kt
/*
 * SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.recorder

import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentFilter
import android.content.ServiceConnection
import android.graphics.drawable.AnimatedVectorDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import android.provider.MediaStore
import android.telephony.TelephonyManager
import android.text.format.DateUtils
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlinx.coroutines.launch
import org.lineageos.recorder.models.UiStatus
import org.lineageos.recorder.service.SoundRecorderService
import org.lineageos.recorder.ui.WaveFormView
import org.lineageos.recorder.utils.LocationHelper
import org.lineageos.recorder.utils.OnBoardingHelper
import org.lineageos.recorder.utils.PermissionManager
import org.lineageos.recorder.utils.PreferencesManager
import org.lineageos.recorder.utils.Utils
import org.lineageos.recorder.viewmodels.RecordingsViewModel
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatterBuilder
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.reflect.safeCast

class RecorderActivity : AppCompatActivity(R.layout.activity_main) {
    // View models
    private val model: RecordingsViewModel by viewModels()

    // Views
    private val contentView by lazy { findViewById<View>(android.R.id.content) }
    private val elapsedTimeText by lazy { findViewById<TextView>(R.id.elapsedTimeTextView) }
    private val floatingActionButton by lazy { findViewById<FloatingActionButton>(R.id.floatingActionButton) }
    private val openSoundListImageView by lazy { findViewById<ImageView>(R.id.openSoundListImageView) }
    private val pauseResumeImageView by lazy { findViewById<ImageView>(R.id.pauseResumeImageView) }
    private val recordingWaveFormView by lazy { findViewById<WaveFormView>(R.id.recordingWaveFormView) }
    private val settingsImageView by lazy { findViewById<ImageView>(R.id.settingsImageView) }
    private val titleTextView by lazy { findViewById<TextView>(R.id.titleTextView) }

    private val locationHelper by lazy { LocationHelper(this) }
    private val permissionManager by lazy { PermissionManager(this) }
    private val preferencesManager by lazy { PreferencesManager(this) }

    private var returnAudio = false
    private var hasRecordedAudio = false

    private var uiStatus = UiStatus.READY

    private val telephonyReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (TelephonyManager.ACTION_PHONE_STATE_CHANGED == intent.action) {
                val state = intent.getStringExtra(TelephonyManager.EXTRA_STATE)
                if (TelephonyManager.EXTRA_STATE_OFFHOOK == state) {
                    togglePause()
                }
            }
        }
    }

    private val messenger = Messenger(object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                SoundRecorderService.MSG_UI_STATUS -> setStatus(msg.obj as UiStatus)
                SoundRecorderService.MSG_SOUND_AMPLITUDE -> setVisualizerAmplitude(msg.obj as Int)
                SoundRecorderService.MSG_TIME_ELAPSED -> setElapsedTime(msg.obj as Long)
                else -> super.handleMessage(msg)
            }
        }
    })

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val messengerService = Messenger(service)
            this@RecorderActivity.service = messengerService
            try {
                val msg = Message.obtain(
                    null,
                    SoundRecorderService.MSG_REGISTER_CLIENT
                )
                msg.replyTo = messenger
                messengerService.send(msg)
            } catch (e: RemoteException) {
                // In this case the service has crashed before we could even
                // do anything with it; we can count on soon being
                // disconnected (and then reconnected if it can be restarted)
                // so there is no need to do anything here.
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {
            service = null
        }
    }
    private var service: Messenger? = null
    private var isServiceBound = false

    public override fun onCreate(savedInstance: Bundle?) {
        super.onCreate(savedInstance)

        // Setup edge-to-edge
        WindowCompat.setDecorFitsSystemWindows(window, false)

        floatingActionButton.setOnClickListener { toggleSoundRecorder() }
        pauseResumeImageView.setOnClickListener { togglePause() }
        openSoundListImageView.setOnClickListener { openList() }
        settingsImageView.setOnClickListener { openSettings() }

        ViewCompat.setOnApplyWindowInsetsListener(contentView) { _, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            contentView.updatePadding(
                bottom = insets.bottom,
            )

            windowInsets
        }

        if (MediaStore.Audio.Media.RECORD_SOUND_ACTION == intent.action) {
            returnAudio = true
            openSoundListImageView.isVisible = false
            settingsImageView.isVisible = false
        }

        doBindService()

        OnBoardingHelper.onBoardList(this, openSoundListImageView)
        OnBoardingHelper.onBoardSettings(this, settingsImageView)
    }

    public override fun onDestroy() {
        doUnbindService()
        super.onDestroy()
    }

    override fun onStart() {
        super.onStart()

        ContextCompat.registerReceiver(
            this,
            telephonyReceiver,
            IntentFilter(TelephonyManager.ACTION_PHONE_STATE_CHANGED),
            ContextCompat.RECEIVER_NOT_EXPORTED
        )
    }

    override fun onStop() {
        unregisterReceiver(telephonyReceiver)

        super.onStop()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>,
        results: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, results)

        if (requestCode == PermissionManager.REQUEST_CODE) {
            if (permissionManager.hasEssentialPermissions()) {
                toggleAfterPermissionRequest()
            } else {
                permissionManager.onEssentialPermissionsDenied()
            }
        }
    }

    private fun setStatus(status: UiStatus) {
        applyUiStatus(status)

        if (returnAudio && hasRecordedAudio) {
            Utils.cancelShareNotification(this)
            promptUser()
        }
    }

    private fun setVisualizerAmplitude(amplitude: Int) {
        recordingWaveFormView.post { recordingWaveFormView.setAmplitude(amplitude) }
    }

    private fun setElapsedTime(seconds: Long) {
        elapsedTimeText.post {
            elapsedTimeText.text = DateUtils.formatElapsedTime(seconds)
        }
    }

    private fun toggleAfterPermissionRequest() {
        Handler(Looper.getMainLooper()).postDelayed({ toggleSoundRecorder() }, 500)
    }

    private fun toggleSoundRecorder() {
        if (permissionManager.requestEssentialPermissions()) {
            return
        }

        if (uiStatus == UiStatus.READY) {
            // Start
            startService(
                Intent(this, SoundRecorderService::class.java)
                    .setAction(SoundRecorderService.ACTION_START)
                    .putExtra(SoundRecorderService.EXTRA_FILE_NAME, newRecordFileName)
            )
        } else {
            // Stop
            startService(
                Intent(this, SoundRecorderService::class.java)
                    .setAction(SoundRecorderService.ACTION_STOP)
            )
            hasRecordedAudio = true
        }
    }

    private fun togglePause() {
        when (uiStatus) {
            UiStatus.RECORDING -> startService(
                Intent(this, SoundRecorderService::class.java)
                    .setAction(SoundRecorderService.ACTION_PAUSE)
            )

            UiStatus.PAUSED -> startService(
                Intent(this, SoundRecorderService::class.java)
                    .setAction(SoundRecorderService.ACTION_RESUME)
            )

            UiStatus.READY -> {}
        }
    }

    private fun applyUiStatus(status: UiStatus) {
        uiStatus = status
        if (UiStatus.READY == status) {
            titleTextView.text = getString(R.string.main_sound_action)
            floatingActionButton.setImageResource(R.drawable.ic_mic)
            elapsedTimeText.isVisible = false
            recordingWaveFormView.isVisible = false
            pauseResumeImageView.isVisible = false
        } else {
            floatingActionButton.setImageResource(R.drawable.ic_stop)
            elapsedTimeText.isVisible = true
            recordingWaveFormView.isVisible = true
            recordingWaveFormView.setAmplitude(0)
            pauseResumeImageView.isVisible = true
            val prDrawable: Drawable?
            if (UiStatus.PAUSED == status) {
                titleTextView.text = getString(R.string.sound_recording_title_paused)
                pauseResumeImageView.contentDescription = getString(R.string.resume)
                prDrawable = ContextCompat.getDrawable(this, R.drawable.avd_play_to_pause)
            } else {
                titleTextView.text = getString(R.string.sound_recording_title_working)
                pauseResumeImageView.contentDescription = getString(R.string.pause)
                prDrawable = ContextCompat.getDrawable(this, R.drawable.avd_pause_to_play)
            }
            pauseResumeImageView.tooltipText = pauseResumeImageView.contentDescription
            pauseResumeImageView.setImageDrawable(prDrawable)
            AnimatedVectorDrawable::class.safeCast(prDrawable)?.start()
        }
    }

    private fun doBindService() {
        bindService(
            Intent(this, SoundRecorderService::class.java),
            serviceConnection, BIND_AUTO_CREATE
        )
        isServiceBound = true
    }

    private fun doUnbindService() {
        service?.let {
            try {
                val msg = Message.obtain(
                    null,
                    SoundRecorderService.MSG_UNREGISTER_CLIENT
                )
                msg.replyTo = messenger
                it.send(msg)
            } catch (e: RemoteException) {
                // There is nothing special we need to do if the service
                // has crashed.
            }
        }

        if (isServiceBound) {
            unbindService(serviceConnection)
            isServiceBound = false
        }
    }

    private fun openSettings() {
        val intent = Intent(this, DialogActivity::class.java)
            .putExtra(
                DialogActivity.EXTRA_IS_RECORDING,
                uiStatus != UiStatus.READY
            )
        startActivity(intent)
    }

    private fun openList() {
        startActivity(Intent(this, ListActivity::class.java))
    }

    private fun confirmLastResult() {
        val resultIntent = Intent().setData(preferencesManager.lastItemUri)
        setResult(RESULT_OK, resultIntent)
        finish()
    }

    private fun discardLastResult() {
        preferencesManager.lastItemUri?.let {
            lifecycleScope.launch {
                model.deleteRecordings(it)
                preferencesManager.lastItemUri = null
            }
        }
        cancelResult(true)
    }

    private fun cancelResult(quit: Boolean) {
        setResult(RESULT_CANCELED, Intent())
        hasRecordedAudio = false
        if (quit) {
            finish()
        }
    }

    private fun promptUser() {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.confirm_result_title)
            .setMessage(R.string.confirm_result_message)
            .setPositiveButton(R.string.confirm) { _: DialogInterface?, _: Int ->
                confirmLastResult()
            }
            .setNegativeButton(R.string.discard) { _: DialogInterface?, _: Int ->
                discardLastResult()
            }
            .setNeutralButton(R.string.record_again) { _: DialogInterface?, _: Int ->
                cancelResult(false)
            }
            .setCancelable(false)
            .show()
    }

    private val newRecordFileName: String
        get() {
            val tag = locationHelper.currentLocationName ?: FILE_NAME_FALLBACK
            val formatter = DateTimeFormatterBuilder()
                .append(DateTimeFormatter.ISO_LOCAL_DATE)
                .appendLiteral(' ')
                .append(DateTimeFormatter.ISO_LOCAL_TIME)
                .toFormatter(Locale.getDefault())
            val now = LocalDateTime.now()
            return String.format(
                FILE_NAME_BASE, tag,
                formatter.format(now.truncatedTo(ChronoUnit.SECONDS))
            ) + ".%1\$s"
        }

    companion object {
        private const val FILE_NAME_BASE = "%1\$s (%2\$s)"
        private const val FILE_NAME_FALLBACK = "Sound record"
    }
}
